### PR TITLE
Improved test coverage of Python code

### PR DESCRIPTION
**Pull request by gedeck**
Following modules were reviewed and coverage improved:

- rdkit.Chem.AtomPairs (95%)
- rdkit.Chem.EState (96%)
- rdkit.Chem.FeatMaps (96%)

The unittests were extended to run the Doctests and thereby include
coverage information for these as well.

### CODE REVIEW

File: rdkit/Chem/AtomPairs/Sheridan.py [hunk 1]
```diff
@@ -13,26 +13,24 @@
...
 
-import os.path, re
-from rdkit import RDConfig
-
 
 def _readPattyDefs(fname=os.path.join(RDConfig.RDDataDir, 'SmartsLib', 'patty_rules.txt')):
   with open(fname, 'r') as inf:
     lines = [x.strip().split('# ')[0].strip() for x in inf]
-  splitl = [re.split('[ ]*', x) for x in lines if x != '']
+  splitl = [re.split('[ ]+', x) for x in lines if x != '']
```
Comment (gedeck) [inline rdkit/Chem/AtomPairs/Sheridan.py]: This change addresses a warning that is now raised in Python 3.5 and will in future releases cause an exception. The old regex matches an empty string. Changing the * to + requires at least one space.

File: rdkit/Chem/EState/EState_VSA.py [hunk 2]
```diff
@@ -71,48 +75,59 @@ def EState_VSA_(mol, bins=None, force=1):
...
+  fn = EState_VSA_bin
+  fn.__name__ = name
+  if hasattr(fn, '__qualname__'):
+    fn.__qualname__ = name
+  fn.__doc__ = _descriptorDocstring('EState VSA', nbin, estateBins)
+  fn.version = '1.0.1'
+  return name, fn
+
+
 def _InstallDescriptors():
```
Comment (gedeck) [inline rdkit/Chem/EState/EState_VSA.py]: I found this code very hard to understand and made the function declaration more explicit.

File: rdkit/TestRunner.py [hunk 3]
```diff
@@ -184,3 +184,40 @@ def ReportResults(script, failedTests, nTests, runTime, verbose, dest):
     else:
       sys.stderr.write('  All %d tests (in %d modules) passed\n' % (totNumRun, len(extras)))
   sys.exit(totNumFailed)
+
+
+class _RedirectStream:
```
Comment (gedeck) [inline rdkit/TestRunner.py]: This is code from Python 3 to get the functionality in Python 2 tests.

**Comment by bp-kelley** [comment]
@gedeck  There's a python2.7 failure here
```
  File "UnitTestEState.py", line 132, in test_exampleCode
    EState.EState._exampleCode()
  File "/home/travis/build/rdkit/rdkit/rdkit/Chem/EState/EState.py", line 110, in _exampleCode
    print(smi)
TypeError: unicode argument expected, got 'str'
```

**Comment by bp-kelley** [comment]
Two more:

```
Traceback (most recent call last):
  File "UnitTestFingerprints.py", line 55, in test_exampleCode
    Fingerprinter._exampleCode()
  File "/home/travis/build/rdkit/rdkit/rdkit/Chem/EState/Fingerprinter.py", line 55, in _exampleCode
    print(smi, Chem.MolToSmiles(m))
TypeError: unicode argument expected, got 'str'
```

```
Traceback (most recent call last):
  File "UnitTestFeatMapParser.py", line 76, in test_FeatMapParser
    with io.StringIO(self.data) as file:
TypeError: initial_value must be unicode or None, not str
```

I find these odd, any idea what is going on here?

**Comment by bp-kelley** [comment]
It looks like in the context of the exec, stdout may be replaced with io.StringIO which only accepts unicode.  This looks fairly nasty to fix:

http://stackoverflow.com/questions/3423601/python-2-7-exec-what-is-wrong

Note the comments at the bottom...

https://gist.github.com/flowerhack/e9747a7b371986e27b26

**Comment by rvianello** [comment]
in my experience the StringIO/ByteIO classes from the six package provide a much better portability across python 2 and 3 compared to the io module implementation (even if io was added to python2 with the intention of providing functionalities similar to the original python3 version).

**Comment by gedeck** [comment]
Replaced with six.StringIO and updated my branch. Looks like the PR is updated too, we'll see if it runs now.

**Comment by bp-kelley** [comment]
Fixed most of them but added one more:
```
Traceback (most recent call last):
  File "UnitTestFeatMapParser.py", line 76, in test_FeatMapParser
    with StringIO(self.data) as file:
AttributeError: StringIO instance has no attribute '__exit__'
```

**Comment by rvianello** [comment]
the py2 StringIO.StringIO implementation doesn't implement the context manager protocol. what about not bothering about closing the StringIO instance in the test code?

**Comment by rvianello** [comment]
or explicitly call close().

**Comment by gedeck** [comment]
I use contextlib.closing around the StringIO(self.data) to ensure that StringIO is closed. Works on Py3 this way and should fix the Python2 problem.

**Comment by bp-kelley** [comment]
@greglandrum I'm happy with this if you want to merge.

File: rdkit/Chem/AtomPairs/Pairs.py [hunk 4]
```diff
@@ -16,14 +16,13 @@
...
 """
-from rdkit.DataStructs import IntSparseIntVect
-from rdkit import Chem
+from rdkit import DataStructs
 from rdkit.Chem import rdMolDescriptors
 from rdkit.Chem.AtomPairs import Utils
-from rdkit import DataStructs
 
-from rdkit.Chem.rdMolDescriptors import GetAtomPairFingerprint, GetHashedAtomPairFingerprint
+GetAtomPairFingerprint = rdMolDescriptors.GetAtomPairFingerprint
```
Comment (greglandrum) [inline rdkit/Chem/AtomPairs/Pairs.py]: Though I have no particular objection to this change, I wonder why you made it.
Is the `from x import y` pattern considered bad now?

File: rdkit/Chem/AtomPairs/UnitTestDescriptors.py [hunk 5]
```diff
@@ -28,47 +40,59 @@ def testPairsRegression(self):
...
-      #    else:
-      #      print '>>>2',k,v
-      #  for k,v in rd.iteritems():
-      #    if pd.has_key(k):
-      #      if pd[k]!=v: print '>>>3',k,v,pd[k]
-      #    else:
-      #      print '>>>4',k,v
-      self.assertTrue(ap == atomPairs[i])
-      self.assertTrue(ap != atomPairs[i - 1])
+      if ap != atomPairs[i]:  # pragma: nocover
```
Comment (greglandrum) [inline rdkit/Chem/AtomPairs/UnitTestDescriptors.py]: I think all this debugging code should be commented back out again.

File: rdkit/Chem/EState/UnitTestVSA.py [hunk 6]
```diff
@@ -12,52 +12,44 @@
...
 from rdkit import Chem
 from rdkit.Chem.EState import EState_VSA
-import os.path
 
 
-def feq(n1, n2, tol=1e-4):
-  return abs(n1 - n2) <= tol
+class TestCase(unittest.TestCase):
 
+  @staticmethod
```
Comment (greglandrum) [inline rdkit/Chem/EState/UnitTestVSA.py]: I don't understand at all why you made this change.
It doesn't seem like a simplification or clarification.
What's the motivation?

File: rdkit/Chem/AtomPairs/Pairs.py (hunk 4 above)
Comment (gedeck) [inline rdkit/Chem/AtomPairs/Pairs.py]: The intention was to highlight the main exports from the function. I changed this back and added a comment instead.

File: rdkit/Chem/AtomPairs/UnitTestDescriptors.py (hunk 5 above)
Comment (gedeck) [inline rdkit/Chem/AtomPairs/UnitTestDescriptors.py]: I moved the debugging code (identical for both fingerprints) into a separate function. That makes it more obvious that this is for information only.

File: rdkit/Chem/EState/UnitTestVSA.py (hunk 6 above)
Comment (bp-kelley) [inline rdkit/Chem/EState/UnitTestVSA.py]: That's a technical distinction.  However python has no real "private static method" and one may reasonably assume that the method is usable without the rest of the class which I don't really think is the case here.  Not that it really matters in this case and it is a purely stylistic concept but I tend to prefer static methods to define external api points, not internal ones.

Comment (greglandrum) [inline rdkit/Chem/EState/UnitTestVSA.py]: Sorry I was unclear in where I put the comment.
I understand what a staticmethod is. I was wondering why you pulled the code to get the reference data out of the single method in which it was used and put it in a separate method.

Comment (gedeck) [inline rdkit/Chem/EState/UnitTestVSA.py]: When I looked at this test it took me a while to figure out what it was doing. Understanding required knowing a few things about how the rest of the code works. Why do we get the expected values from the molecule, how do we access the functions?

By separating the loading of the validation data (that could be a better name) from the comparison it seemed to me clearer. Looking at it again, i think the population of the property function dictionary should be moved to the test for a clearer separation.

Comment (greglandrum) [inline rdkit/Chem/EState/UnitTestVSA.py]: I can live with that. Accepting the PR

**Pull request merged**